src/redisearch_rs/slots_tracker/src/lib.rs

```diff
@@ -0,0 +1,374 @@
+fn sync_version(tracker: &SlotsTracker) {
+    STATE
+        .version
+        .store(tracker.get_version(), Ordering::Relaxed);
+}
+
+// ============================================================================
+// C FFI Interface - Public API
+// ============================================================================
+
+/// C-compatible slot range structure matching RedisModuleSlotRange.
+#[repr(C)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct SlotRange {
+    pub start: u16,
+    pub end: u16,
+}
+
+/// C-compatible slot range array structure matching RedisModuleSlotRangeArray.
+///
+/// This is a variable-length structure with a flexible array member.
+#[repr(C)]
+pub struct SlotRangeArray {
+    pub num_ranges: i32,
+    pub ranges: [SlotRange; 0], // Flexible array member
+}
+
+// ============================================================================
+// Private Helper Functions
+// ============================================================================
+
+/// Converts a C SlotRangeArray pointer to a Rust slice.
+///
+/// # Panics
+///
+/// Panics if the pointer is null or num_ranges is less than 0.
+///
+/// # Safety
+///
+/// The caller must ensure the pointer is valid and points to a properly initialized
+/// SlotRangeArray with at least `num_ranges` elements in the flexible array.
+unsafe fn parse_slot_ranges<'a>(ranges: *const SlotRangeArray) -> &'a [SlotRange] {
+    assert!(!ranges.is_null(), "SlotRangeArray pointer is null");
+
+    // SAFETY: Caller guarantees valid pointer
+    let ranges = unsafe { &*ranges };
+
+    assert!(
+        ranges.num_ranges >= 0,
+        "num_ranges must be at least 0, got {}",
+        ranges.num_ranges
+    );
+
+    // SAFETY: Caller guarantees the flexible array has num_ranges elements
+    unsafe { std::slice::from_raw_parts(ranges.ranges.as_ptr(), ranges.num_ranges as usize) }
+}
+
+// ============================================================================
+// Main API Functions
+// ============================================================================
+
+/// Sets the local responsibility slot ranges.
+///
+/// This function updates the "local slots" set to match the provided ranges.
+/// If the ranges differ from the current configuration:
+/// - Updates "local slots" to the new ranges
+/// - Removes any overlapping slots from "fully available slots" and "partially available slots"
+/// - Increments the version counter
+///
+/// If the ranges are identical to the current configuration, no changes are made.
+///
+/// # Safety
+///
+/// This function must be called from the main thread only.
+/// The `ranges` pointer must be valid and point to a properly initialized RedisModuleSlotRangeArray.
+/// The ranges array must contain `num_ranges` valid elements.
+/// All ranges must be sorted and have start <= end, with values in [0, 16383].
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn slots_tracker_set_local_slots(ranges: *const SlotRangeArray) {
```

Comment:
You'll want the C -> Rust API layer to be in the `c_entrypoint` directory. This means making another crate there with the API you want. You can just follow the structure of one of the existing crates in that directory. And will want the `cbindgen` to generate the C headers for you.

With that change, this crate should have no `unsafe` code and should be a pure Rust library.